src/http/api/resources/dag.js

```diff
@@ -0,0 +1,237 @@
+    } catch (err) {
+      throw Boom.boomify(err, { message: 'Failed to get dag node' })
+    }
+
+    if (key.codec === 'dag-pb' && result.value) {
+      if (typeof result.value.toJSON === 'function') {
+        result.value = result.value.toJSON()
+      }
+
+      if (Buffer.isBuffer(result.value.data)) {
+        result.value.data = result.value.data.toString(request.query.dataencoding)
+      }
+    }
+
+    return h.response(result.value)
+  }
+}
+
+exports.put = {
+  validate: {
+    query: Joi.object().keys({
+      // TODO validate format, & hash
+      format: Joi.string(),
+      'input-enc': Joi.string().valid('dag-cbor', 'dag-pb', 'raw'),
+      pin: Joi.boolean(),
+      hash: Joi.string(),
+      'cid-base': Joi.string().valid(multibase.names).default('base58btc')
+    }).unknown()
+  },
+
+  // pre request handler that parses the args and returns `node`
+  // which is assigned to `request.pre.args`
+  async parseArgs (request, h) {
+    if (!request.payload) {
+      throw Boom.badRequest("File argument 'data' is required")
+    }
+
+    const enc = request.query.inputenc
+
+    const fileStream = await new Promise((resolve, reject) => {
+      multipart.reqParser(request.payload)
+        .on('file', (name, stream) => resolve(stream))
+        .on('end', () => reject(Boom.badRequest("File argument 'data' is required")))
+    })
+
+    let data = await new Promise((resolve, reject) => {
+      fileStream
+        .on('data', data => resolve(data))
+        .on('end', () => reject(Boom.badRequest("File argument 'data' is required")))
+    })
+
+    if (enc === 'json') {
+      try {
+        data = JSON.parse(data.toString())
+      } catch (err) {
+        throw Boom.badRequest('Failed to parse the JSON: ' + err)
+      }
+    }
+
+    try {
+      return {
+        buffer: data
+      }
+    } catch (err) {
+      throw Boom.badRequest('Failed to create DAG node: ' + err)
+    }
+  },
+
+  // main route handler which is called after the above `parseArgs`, but only if the args were valid
+  async handler (request, h) {
+    const { ipfs } = request.server.app
+    const { buffer } = request.pre.args
+
+    let cid
+
+    return new Promise((resolve, reject) => {
+      const format = ipfs._ipld.resolvers[request.query.format]
+
+      if (!format) {
+        return reject(Boom.badRequest(`Missing IPLD format "${request.query.format}"`))
+      }
+
+      format.util.deserialize(buffer, async (err, node) => {
+        if (err) {
+          return reject(err)
+        }
+
+        try {
+          cid = await ipfs.dag.put(node, {
+            format: request.query.format,
+            hashAlg: request.query.hash
+          })
+        } catch (err) {
+          throw Boom.boomify(err, { message: 'Failed to put node' })
+        }
+
+        if (request.query.pin) {
+          await ipfs.pin.add(cid)
+        }
+
+        resolve(h.response({
+          Cid: {
+            '/': cid.toBaseEncodedString(request.query.cidbase)
```

Comment:
Please use the [`cidToString`](https://github.com/ipfs/js-ipfs/blob/master/src/utils/cid.js) utility so this doesn't throw and auto upgrades the CID to v1 when appropriate.

e.g. `cidToString(cid, { base: request.query['cid-base'] })`